pkg/config/expand.go

```diff
@@ -166,6 +169,48 @@ func (bp Blueprint) expandBackend(grp *Group) {
 	}
 }
 
+func getDefaultGoogleProviders(bp Blueprint) map[string]TerraformProvider {
+	gglConf := Dict{}
+	for s, v := range map[string]string{
+		"project": "project_id",
+		"region":  "region",
+		"zone":    "zone"} {
+		if bp.Vars.Has(v) {
+			gglConf = gglConf.With(s, GlobalRef(v).AsValue())
+		}
+	}
+	return map[string]TerraformProvider{
+		"google": TerraformProvider{
+			Source:        "hashicorp/google",
+			Version:       ">= 4.84.0, < 5.32.0",
+			Configuration: gglConf},
+		"google-beta": TerraformProvider{
+			Source:        "hashicorp/google-beta",
+			Version:       ">= 4.84.0, < 5.32.0",
+			Configuration: gglConf}}
+}
+
+func (bp Blueprint) expandProviders(grp *Group) {
+	// 1. DEFAULT: use TerraformProviders provider dictionary (if supplied)
+	// 2. If top-level TerraformProviders is defined, insert that
+	//    provider dictionary into resource groups which have no explicit
+	//    TerraformProviders
```

Comment:
I struggle to understand motivation behind treating `bp.TerraformProviders` and `getDefaultGoogleProviders(bp)` differently:
* `getDefaultGoogleProviders(bp)` - all or nothing;
* `bp.TerraformProviders` - "join" with `grp.TerraformProviders`.

Consider simple "all or nothing" alternative:

```golang
default := bp.Providers
if default == nil { default = getDefaultGoogleProviders(bp) }

if grp.Providers == nil { grp.Providers = default }
```

Also it's important to diffentiate `providers == nil` and `len(providers) == 0`. What if user intentionally specified `terraform_providers: {}` to omit all of them without intention to fallback to "defaults".

Reply:
Understood, I fixed it.  Yeah I think I was trying to be too fancy with the original logic and didn't consider the "no provider" path.